docs/auth_chain_difference_algorithm.md

```diff
@@ -0,0 +1,78 @@
+# Auth Chain Difference Algorithm
+
+The auth chain difference algorithm is used by V2 state resolution, where a
+naive implementation can be a significant source of CPU and DB usage.
+
+The auth chain difference of a set of state sets is the union minus the
+intersection of the sets of auth chains corresponding to the state sets, i.e an
+event is in the auth chain difference if it is reachable by walking the auth
+event graph from at least one of the state sets but not from *all* of the state
+sets.
+
+## Chain Cover Index
+
+Synapse computes auth chain differences by pre-computing a "chain cover" index
+for the auth chain in a room, allowing efficient reachability queries like "is
+event A in the auth chain of event B". This is done by assigning every event a
+*chain ID* and *sequence number* and having map of *links* such that A is
+reachable by B (i.e. `A` is in the auth chain of `B`) if and only if either:
+
+1. A and B have the same chain ID and `A`'s sequence number is less than `B`'s
+   sequence number; or
+2. there is a link `L` between `B`'s chain ID and `A`'s chain ID such that
+   `L.seq_no` <= `B.seq_no` and `A.seq_no` <= `L.seq_no`.
+
+There are actually two variants, one where we store links from each chain to
+every other reachable chain (the transitive closure of the links graph), and one
+where we remove redundant links (the transitive reduction of the links graph)
+e.g. if we have chains `C3 -> C2 -> C1` then the link `C3 -> C1` would not be
+stored. Synapse uses the former variant so that it doesn't need to recurse to
+test reachability between chains.
+
+### Example
+
+An example auth graph would look like the following, where chains have been
+formed based on type/state_key and are denoted by colour and are labelled with
+`(chain ID, sequence number)`. Links are denoted by the arrows (links in grey
+are those that would be remove in the second variant described above).
+
+![Example](auth_chain_diff.dot.png)
+
+Note that we don't add links between every event and its auth events, as that is
+redundant (under both variants), e.g. all events point to the create event, but
+each chain only needs the one link from it's base to the create event.
```

Comment:
This paragraph is dense. In particular "but each chain only needs" seems a bit confusing, I think leaving out the "but" would simplify it.